File: leetcode_75/Binary_Tree_BFS/python/1161.py
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        result = [-10 ** 6, 0]
        q = deque([root])
        k = 0
        while q:
            k += 1
            s = 0
            for _ in range(len(q)):
                node = q.popleft()
                s += node.val

                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            if s > result[0]:
                result[0] = s
                result[1] = k

        return result[1]

File: leetcode_75/Binary_Tree_BFS/cpp/1161.cpp
#include <deque>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int maxLevelSum(TreeNode *root)
    {
        int mx_val = -1000000;
        int level = 0;

        int k = 0;
        int s, q_size;
        deque<TreeNode *> q = {root};
        TreeNode *node = nullptr;
        while (!q.empty())
        {
            k++;
            s = 0;
            q_size = q.size();
            for (int i = 0; i < q_size; ++i)
            {
                node = q.front();
                s += node->val;
                q.pop_front();

                if (node->left)
                {
                    q.push_back(node->left);
                }
                if (node->right)
                {
                    q.push_back(node->right);
                }
            }
            if (s > mx_val)
            {
                mx_val = s;
                level = k;
            }
        }

        return level;
    }
};